File: server/src/test/java/com/fluffy/support/testcontainers/RedisContainerInitializer.java
package com.fluffy.support.testcontainers;

import java.util.Map;
import org.jetbrains.annotations.NotNull;
import org.springframework.boot.test.util.TestPropertyValues;
import org.springframework.context.ApplicationContextInitializer;
import org.springframework.context.ConfigurableApplicationContext;
import org.testcontainers.containers.GenericContainer;

public class RedisContainerInitializer implements ApplicationContextInitializer<ConfigurableApplicationContext> {

    private static final GenericContainer<?> REDIS = new GenericContainer<>("valkey/valkey:8-alpine")
            .withExposedPorts(6379);

    static {
        REDIS.start();
    }

    @Override
    public void initialize(@NotNull ConfigurableApplicationContext context) {
        Map<String, String> properties = Map.of(
                "spring.data.redis.host", REDIS.getHost(),
                "spring.data.redis.port", String.valueOf(REDIS.getFirstMappedPort()),
                "spring.data.redis.ssl.enabled", "false"
        );

        TestPropertyValues.of(properties).applyTo(context);
    }
}

File: server/src/test/kotlin/com/fluffy/support/testcontainers/RedisContainerInitializer.kt
package com.fluffy.support.testcontainers

import org.springframework.boot.test.util.TestPropertyValues
import org.springframework.context.ApplicationContextInitializer
import org.springframework.context.ConfigurableApplicationContext
import org.testcontainers.containers.GenericContainer

class RedisContainerInitializer : ApplicationContextInitializer<ConfigurableApplicationContext> {

    companion object {
        private val REDIS = GenericContainer("valkey/valkey:8-alpine")
            .withExposedPorts(6379)

        init {
            REDIS.start()
        }
    }

    override fun initialize(context: ConfigurableApplicationContext) {
        val properties: Map<String, String> = mapOf(
            "spring.data.redis.host" to REDIS.host,
            "spring.data.redis.port" to REDIS.firstMappedPort.toString(),
            "spring.data.redis.ssl.enabled" to false.toString()
        )

        TestPropertyValues.of(properties).applyTo(context)
    }
}